gen/src/generate/idna/mapping.rs

```diff
@@ -0,0 +1,143 @@
+    /// There are two values: NV8 and XV8.
+    /// NV8 is only present if the status is valid but the character is excluded
+    /// by IDNA2008 from all domain names for all versions of Unicode.
+    /// XV8 is present when the character is excluded
+    /// by IDNA2008 for the current version of Unicode.
+    /// These are not normative values.
+    idna_2008_status: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct IdnaMapping(Box<[IdnaDataLine]>);
+
+impl IdnaMapping {
+    fn emit<P: AsRef<Path>>(&self, dir: P) -> io::Result<()> {
+        let mut file = File::create(dir.as_ref().join("idna_mapping.rsv"))?;
+        let IdnaMapping(ref lines) = *self;
+        writeln!(file, "&[")?;
+        for line in lines.iter() {
+            write!(
+                file,
+                "    ('{}', '{}', {}",
+                line.low.escape_unicode(),
+                line.high.escape_unicode(),
+                match line.status.as_str() {
+                    "valid" => "Valid",
+                    "ignored" => "Ignored",
+                    "mapped" => "Mapped",
+                    "deviation" => "Deviation",
+                    "disallowed" => "Disallowed",
+                    "disallowed_STD3_valid" => "DisallowedStd3Valid",
+                    "disallowed_STD3_mapped" => "DisallowedStd3Mapped",
+                    _ => unreachable!(),
+                },
+            )?;
+            if let Some(ref chars) = line.mapping {
+            write!(file, "(&[")?;
+                for char in chars.iter() {
+                    write!(file, "'{}',", char.escape_unicode())?;
+                }
+            write!(file, "])")?;
+            }
+            writeln!(file, "),")?;
+        }
+        writeln!(file, "]")?;
+        Ok(())
+    }
+}
+
+impl FromStr for IdnaMapping {
+    type Err = ();
+
+    fn from_str(str: &str) -> Result<Self, Self::Err> {
+        lazy_static! {
+            static ref REGEX: Regex = Regex::new(
+                r"(?xm)(?xm)^
+                  ([[:xdigit:]]{4,6})                              # low code point
+                  (?:\.\.([[:xdigit:]]{4,6}))?                     # high code point (option)
+                  [[:space:]]*;[[:space:]]*
+                  ([[:word:]]*)                                    # status
+                  [[:space:]]*
+                  (?:
+                    ;[[:space:]]*
+                    ([[:xdigit:]]{4,6}(?:\x20[[:xdigit:]]{4,6})*)? # mapping (option)
+                    [[:space:]]*
+                    (?:
+                      ;[[:space:]]*
+                      (NV8|XV8)                                    # IDNA2008 Status (option)
+                      [[:space:]]*
+                    )?
+                  )?
+                \#"
+            ).unwrap();
+        }
+
+        /// impl note:
+        /// 8393 is the number of data lines in the Unicode version 10.0.0 IdnaMappingTable.txt
+        let mut entries = Vec::with_capacity(8393);
```

Comment:
I could always `collect` and use the length of that, but in that case there's no point because we grew the first vector in the collect.

Unfortunately, [the `captures_iter` iterator has no special information about the number of matches](https://doc.rust-lang.org/regex/regex/struct.CaptureMatches.html). We could count the number of lines and do a simple estimation that (say) 90% of lines are data lines, but that seems just as fragile.

The best answer may be to pick a nice number like 0x2200 (8704 in decimal) that gives a bit of room to grow, which also has the nice property of being one 128th of the total space of codepoints (0x110000).